core/src/main/java/org/apache/iceberg/TableProperties.java

```diff
@@ -231,6 +231,10 @@ private TableProperties() {}
   public static final String ORC_BATCH_SIZE = "read.orc.vectorization.batch-size";
   public static final int ORC_BATCH_SIZE_DEFAULT = 5000;
 
+  public static final String DELETE_COLUMN_STATS_FILTERING_ENABLED =
+      "read.delete.column-stats-filtering.enabled";
```

Comment:
I am still thinking about the best name. Ideas are welcome.

Reply:
I don't think that we should have an option to do this.

What we talked about was only applying stats filtering if the min and max file are the same, and otherwise disabling it. Why have an option when we know what the setting should be?

Reply:
I reconsidered that a bit after spending more time profiling.

- We want to get rid of any extra overhead (like Stream API) and transforms if stats filtering is disabled. The deletes are being looked up in tight loops and MoR use cases are frequently needed for larger tables where CoW is just too expensive. That means we need to look up deletes for a large number of data files. Any extra work matters. All of this overhead is present in flamegraphs. This also becomes even more critical when data files are loaded in memory (distributed planning).
- We want to skip even loading delete stats if they will not be used.
- We may reconsider how our paths are being generated and sometimes include a temporal part to facilitate filtering. That won't be always possible but we can't assume the current way of generating file names is the only way to do that.
- We want a way to disable column stats filtering for some equality delete use cases where filtering is not beneficial. We don't know how useful stats for equality deletes are. Even if they are different, they may be selective. So we can't skip indexing equality delete stats if they are different but they may be still useless.

Reply:
What are your thoughts, @rdblue?

Reply:
We can keep track whether we found any position delete files with the same min/max boundaries while building the index but other points would still apply.

Reply:
Another use case where it is beneficial to know whether delete stats filtering is on upfront is distributed planning. If stats are on, I need to use `ParallelIterable`. If not, there would be substantial overhead with that.

Reply:
I think I wasn't very clear about what I was trying to say. I'll reply to some of the specific points:

> We want a way to disable column stats filtering for some equality delete use cases where filtering is not beneficial.

I completely agree that we should disable the filtering. What I'm saying is that we should _completely_ disable it and not provide the option to turn it on or off. We don't have an expectation that it is helpful, except in the case where the min and max are identical.

I also think it would be good to detect the case where the path's `min == max` and put those positional delete files in a map. I think that is worth it because we have two cases for a positional delete file:
1. There are deletes for multiple data files and stats aren't going to help
2. There are deletes for just one file and stats tell us that

Those are very different cases because of the number of delete files. If you have a strategy to rewrite position deletes into one file per data file, then you'd have a _lot_ of very narrowly targeted delete files. Using a map keyed by the delete path for this case makes a lot of sense because of the number of data files that are removed from the current delete index entirely.

> We want to skip even loading delete stats if they will not be used

I agree in principle, but I don't think this is a huge cost. The data has to be read into memory anyway, we avoid converting to `String`, and it all runs in parallel.

The biggest cost by far is comparison --- in part because of the expense of comparing large strings, but more related to the number of comparisons. We have to compare each data file against each delete file with a newer sequence number. That amounts to `O(m*n)` comparisons if we have `m` delete files and `n` data files. Reducing `m` by separating out `min == max` delete files and then not running the stats comparison for the remaining position deletes seems like a really good start.

I could be convinced that we need to avoid even loading the maps, but I'd like to be convinced by benchmark numbers. Plus, we still need to read the upper and lower bounds for equality deletes. Not loading stats complicates the projection.

Another idea is to add the ability to load data for a subset of columns by ID.

> We may reconsider how our paths are being generated and sometimes include a temporal part to facilitate filtering

Wouldn't we just be back to loading stats then? If I understand correctly, this would make the stats check more valuable, but it's more expensive than we want to even run it right now.

> We want to get rid of any extra overhead (like Stream API) and transforms if stats filtering is disabled

I agree. Don't we often do this by not adding filters or transforms if they aren't needed?